### PR TITLE
[adhoc] Prepare aws2 ClientConfiguration for json serialization and cleanup AWS Module

### PR DESCRIPTION
**Pull request by mosche**
This prepares ClientConfiguration for Json serialization to be used in PipelineOptions (as preparation for https://issues.apache.org/jira/browse/BEAM-13956).

Removing`SerializableAwsCredentialsProvider` and handling by explicit serialization/deserialisation in ClientConfiguration makes this a bit easier.

Additionally cleanup the AwsModule for Json serialization as it grows quickly:
- Use Annotations where possible instead of writing custom Serializers / Deserializers
- Move Annotations to internal classes where possible rather than having Mixin (and move respective tests accordingly)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/validate-runner/README.md)

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/common/ClientConfiguration.java [hunk 1]
```diff
@@ -132,58 +156,8 @@ public Builder retry(Consumer<RetryConfiguration.Builder> retry) {
...
-    public ClientConfiguration build() {
-      if (credentialsProvider() != null) {
-        credentialsProvider(new SerializableAwsCredentialsProvider(credentialsProvider()));
-      }
-      return autoBuild();
-    }
-  }
-
-  /** Internal serializable {@link AwsCredentialsProvider}. */
-  private static class SerializableAwsCredentialsProvider
```
Comment (mosche) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/common/ClientConfiguration.java:148]: Removing SerializableAwsCredentialsProvider to simplify Json serialization, otherwise this would require a custom serializer (and can't be private any longer). Instead ClientConfiguration just stores the serialized Json representation internally.

**Comment by echauchot** [comment]
@mosche thanks for your contribution ! Reviewing ...

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/options/AwsModuleTest.java [hunk 2]
```diff
@@ -151,33 +149,6 @@ public void testProxyConfigurationSerializationDeserialization() throws Exceptio
     assertEquals("password", deserializedProxyConfiguration.password());
   }
 
-  @Test
```
Comment (echauchot) [inline sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/options/AwsModuleTest.java:154]: Agree, better to move them to dedicated test files

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKeyTest.java [hunk 3]
```diff
@@ -54,4 +55,18 @@ public void buildWithArgs(String key, String algorithm, String md5, String encod
     assertEquals(algorithm, sseCustomerKey.getAlgorithm());
     assertEquals(encodedMD5, sseCustomerKey.getMD5());
   }
+
+  @Test
+  public void testJsonSerializeDeserialize() {
+    // default key created by S3Options.SSECustomerKeyFactory
+    SSECustomerKey emptyKey = SSECustomerKey.builder().build();
+    assertThat(jsonSerializeDeserialize(emptyKey)).isEqualToComparingFieldByField(emptyKey);
```
Comment (echauchot) [inline sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKeyTest.java:63]: I know you just moved the test method, but why is this needed to use `isEqualToComparingFieldByField` rather than `isEqualTo` in that case ?

**Review by echauchot — CHANGES_REQUESTED** [review]
thanks Moritz for this PR! LGTM modulo a minor change and a question

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/ClientConfigurationTest.java [hunk 4]
```diff
@@ -50,4 +52,28 @@ public void testSerialization() {
 
     assertThat(deserializedConfig).isEqualTo(config);
   }
+
+  @Test
+  public void testJsonSerialization() {
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/ClientConfigurationTest.java:57]: Initially I planned to ignore RetryConfiguration when serializing to/from Json. While I don't see a use case yet, it might be misleading / error prone. So added that capability and respective tests ...

`regionId` is purely internal and tested through the respective public api using `region`

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/HttpClientConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            .connectionAcquisitionTimeout(100)
+            .connectionMaxIdleTime(200)
+            .connectionTimeout(300)
+            .connectionTimeToLive(400)
+            .socketTimeout(500)
+            .readTimeout(600)
+            .writeTimeout(700)
+            .maxConnections(10)
+            .build();
+
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/HttpClientConfigurationTest.java:42]: It felt a bit to repetitive with the number of properties here ;)

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKeyTest.java (hunk 3 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKeyTest.java:63]: `SSECustomerKey` lacks an implementation of equals ....

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/ClientConfigurationTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/ClientConfigurationTest.java:57]: @echauchot I've pushed that change

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
So many flaky tests all around :(

**Comment by mosche** [comment]
@echauchot Some unrelated tests keep failing 😫 
Giving it another try ...

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
😡

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
@echauchot tests finally passed

**Comment by mosche** [comment]
@echauchot ping :)

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/ClientConfigurationTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (echauchot) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/ClientConfigurationTest.java:57]: :+1:

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/HttpClientConfigurationTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (echauchot) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/common/HttpClientConfigurationTest.java:42]: yes, indeed

**Review by echauchot — APPROVED** [review]
LGTM ! Thanks Moritz !
